src/toolchain.ml

```diff
@@ -184,6 +184,63 @@ let of_settings () : t option Promise.t =
   | Some Global -> Promise.return (Some Global)
   | Some (Custom template) -> Promise.return (Some (Custom template))
 
+let detect_esy_sandbox ~project_root esy () =
+  let open Promise.Syntax in
+  let* esy = esy in
+  let* esy_build_dir_exists =
+    Fs.exists Path.(project_root / "_esy" |> Path.to_string)
+  in
+  let+ manifest = Esy.find_manifest_in_dir project_root in
```

Comment:
if these operations are independent, they can be run concurrently with `Promise.all`. Or we're running them sequentially, we can avoid the computation of `esy_build_dir_exists` if `esy` is `None`, etc.